### ISSUE TITLE
The experiment setting of comparison with SuperPoint

### ISSUE THREAD

**Issue by triangleCZH** [cvg/SOLD2#12]
Hi, thank you so much for the paper and the code:) I find it very interesting that SOLD2 has almost double the performance of SuperPoint on WireFrame. I have some questions for this comparison, since I have been worked on SuperPoint for a while:
1. Did you train superpoint on wireframe to compare, or just use magicleap's pretrain weight?
2. SuperPoint:SOLD2 = 0.58:0.94 (see Fig. 1). Given that the rotation during validation is union in  -90 ~ 90 degree, did you observe that SuperPoint mainly did bad on data with rotate <-45 or >45 (the half with larger angle)? This assumption is based on the shown graph, where SuperPoint seems to perform poorly on large angle data (see Fig. 2)

Fig. 1
<img width="512" alt="image" src="https://user-images.githubusercontent.com/20054340/122455164-1b8b3600-cf7a-11eb-80df-2c4901eb6108.png">

Fig. 2
<img width="998" alt="image" src="https://user-images.githubusercontent.com/20054340/122455082-044c4880-cf7a-11eb-8da8-b86c0fea9ac7.png">

**Comment by rpautrat** [comment]
Hi,

You are right to ask these questions as this experiment was indeed a bit biased in favor of SOLD².

1. We used the pre-trained weights of MagicLeap for the comparison with SuperPoint, and did not fine-tune it on the Wireframe images. Thus, SOLD² has a first advantage here.
2. SuperPoint was indeed mainly failing for image pairs with more than 45 degrees of rotation. Which makes sense, since SuperPoint has only been trained with rotations in [-45, +45].

So all in all, this experiment was indeed biased and detrimental for SuperPoint. We mainly added the baseline SuperPoint to give a rough feeling of a comparison with a point descriptor, but did not have time to make it fairer. At least, it shows the improved robustness to rotation compared to point features.

**Comment by triangleCZH** [comment]
@rpautrat Thank you for your reply. Then I might want to finetune superpoint on wireframe for further comparison and see. 
Therefore, may I know about:
1. Would you provide evaluation code for wireframe dataset, that could generate structural distance and orthogonal distance as described in the paper? For now, I do not see any explicitly.
2. How to compute metrics in the following figure (see Fig. 1)? Especially for "inliers", for SOLD2 I assume it indicates the inlier  lines, but for superpoint there is no line module, so does it mean the inlier points?

Thank you again for this insightful work. I believe it means a lot in the self-supervised feature learning task.

<img width="529" alt="image" src="https://user-images.githubusercontent.com/20054340/122588509-2fdc3b00-d02d-11eb-824e-7c241a7b7dfb.png">

**Comment by rpautrat** [comment]
Hi,

1. I can provide the evaluation code with the implementation of the metrics and line distances. However this will require some code cleaning first. With CVPR this week and some summer travelling afterwards, I might only release it later in July probably.
2. The homography adaptation task is the same as the one introduced in the [SuperPoint](https://arxiv.org/abs/1712.07629) paper. The number of inliers is in both cases (for lines and points) the number of inliers in the RANSAC-based homography estimation, and Reproj. error is the average reprojection error of these inliers (using the orthogonal line distance for the line case).

**Comment by triangleCZH** [comment]
@rpautrat Thank you for your reply. And I hope you will have a wonderful summer vacation:) I would keep this issue open, if I have questions later about implementing the evaluation script.

**Comment by rpautrat** [comment]
Thank you! Yes, better to keep the issue open for now :)

**Comment by triangleCZH** [comment]
Hi, I plan to implement the evaluation script in order to reproduce the results in your paper, and I have two follow-up questions:

1. If I understand it correctly, the evaluation dataset is similar to HPatch, including mages, fixed homographies and the corresponding warped views. May I know if you will release the fixed set of homographies for the validation data? Without the exactly same setting, it is hard to tell whether the it is reproduced or not, even if I implement the evaluation script.
2. Still following Fig. 1, do you mind explain what is the reprojection error? In SuperPoint, the success/failure of a homography  estimation is determined by the corner point distance with a threshold, does reprojection error mean a same thing?

Thank you!

Fig. 1
<img width="512" alt="image" src="https://user-images.githubusercontent.com/20054340/123341699-6bd73a80-d51c-11eb-9caa-ee04855096fe.png">

**Comment by rpautrat** [comment]
1. You can find the Wireframe test images that we used, together with the homography and reference line segments in the npz files [here](https://www.polybox.ethz.ch/index.php/s/zQwjg3kx2iGSNWe).
2. Reprojection error is the reprojection error in RANSAC. For both lines and points we run RANSAC to get a homography from the feature matches (the OpenCV findHomography for points, and a custom-made RANSAC for lines). Reprojection error is thus the average distance of the inliers using the estimated transform.

**Comment by triangleCZH** [comment]
Hi, I have implemented the detection metrics and get slight difference, and I guess it comes from one of the following:
1. For repeatability, in the paper it says 'This metric is computed symmetrically across the two images and averaged'. Does it mean that: for image pair A, B, you warp A's lines to B' space, compute result, then swap A, B to do again, and take average of the two? i.e. ((A_lines in B) / A_lines + (B_lines in A) / B_lines) / 2
2. For localization error, do you also do in a symmetric way just like repeatability?
3. Do you filter out invalid line segs before evaluating? i.e., some lines after warping are out of the image, and might not be considered, I notice this is not done in export_line_features.py.

**Comment by rpautrat** [comment]
Hi,

As far as I remember (I am on the go now and without access to the code):
1. We always only warp the lines of image B into the reference image A (ie A is the original image, B is warped). This ensures that the pixel error threshold is not skewed by the homography (which would happen if we would compare the lines in image B). Then repeatability is defined like in Superpoint: (A_lines in B + B_lines in A) / (A_lines + B_lines).
2. Yes.
3. Yes we only consider lines that are shared between the two views. We might also remove very short lines as well, I would need to check the code.

**Comment by brade31919** [comment]
Hi @triangleCZH ,

1. Yes, as @rpautrat described, we only warp the lines in the warped image (B) back to the ref image (A). Having these two sets of line segments (both in the coordinate system of ref image), then you can compute the repeatability following eq 14~15 in our supplementary material.
2. Yes.
3. We will remove all the line segments that are outside the image border after the warping. We didn't remove very short line segments. We remove all the line segments with junctions having x, y < 0 or > W-1, H-1 (H and W are the height and width of the image), but we won't remove the junctions that lie exactly on the image border (0, W-1, and H-1). Otherwise, you'll end up ignoring lots of long segments delineating the room layout.

Best

**Comment by triangleCZH** [comment]
@brade31919 
Thank you for your explanation. I think the detection metrics is very clear now:)
Do you mind if I ask a little bit more about description metrics details? They will really help people understand the procedure before the release of the official code:)

1. I notice that you use precision and recall both with ground truth and with predicted lines. Are Table 2 and Figure 6 (also the one on your github repo README.md) in your paper made with predicted lines?
2. If so, may I know how do you compute the precision and recall for predicted lines? Because in this case we do not have "True positives" and "True negatives"
3. May I know if the above metrics is done, following the matching result of `line_matcher`, or is it following the result of `line_detection`?
4. Similar to 3., when you estimate the homography and compare with SuperPoint,  is the ransac done after the `line_matcher`?
<img width="693" alt="image" src="https://user-images.githubusercontent.com/20054340/123564345-210d2b00-d787-11eb-9d3c-e5d3b08e570d.png">
<img width="734" alt="image" src="https://user-images.githubusercontent.com/20054340/123564341-19e61d00-d787-11eb-96db-c0278021e3c0.png">

**Comment by triangleCZH** [comment]
Additionally, may I know about the pipeline how you get the final result? (I could open a new issue if you prefer)
Currently I am doing 200 epoch with detector + 20 epoch with paired_desc, but the 200 epoch detector result does not look quite good. Would you suggest using the synthetic pretrain (which labels the dataset) to train detector?

**Comment by brade31919** [comment]
Hi @triangleCZH ,

Sorry for the late reply!

In the wireframe dataset, the synthetic warps (homographies) are sampled so we know the warping, and similarly we know the camera poses in ETH 3D dataset. Given the 2D warping (and 3D transformation), we can compared the detected line segments in the 2 views, and we can use the structural distance (and euclidean distance) to define the correspondences. For example, as mentioned in the caption of Table 2, we use the line segment pairs with structural distances < 5 as positive pairs.

Yes, the results are from `line_matcher` (the predictions). You can't compute the description metrics using only detection results.

Yes, the line matcher will give the the matched pairs in the two views. Then, you can use RANSAC to perform the outlier rejection.

**Comment by brade31919** [comment]
Hi @triangleCZH ,

Can you provide more details about the pipeline you are using currently? Typically, we used the model trained on the synthetic dataset to generate the pseudo ground truths on the real-world dataset (the wireframe dataset) for us. Then, when training the model on the wireframe dataset, you can use the weights of the synthetic pretrained model to initialize your model. This will make the training process more stable.

Can you share some results from your trained detector?

**Comment by triangleCZH** [comment]
Hi @brade31919 
Thank you for your reply!
1. For correspondence evaluation, do you use the detected lines (before match), or the matched lines? From my understanding of your comment, you are using the detected lines, so that the picked lines (dist < 5) are same to the ones picked in repeatability calculation. 

For the second comment, I will update when my new training is finished. Thank you!

**Comment by rpautrat** [comment]
Hi @triangleCZH,

Regarding the correspondence evaluation, you are correct: we use as ground truth line correspondences, the detected lines which are repeatable (dist < 5), similarly as in the repeatability evaluation.
